src/server/src/main/java/io/cassandrareaper/service/Heart.java

```diff
@@ -83,7 +94,9 @@ synchronized void beat() {
   synchronized void beatMetrics() {
     if (context.storage instanceof IDistributedStorage
             && ReaperApplicationConfiguration.DatacenterAvailability.EACH
-                == context.config.getDatacenterAvailability()) {
+                == context.config.getDatacenterAvailability()
+        || ReaperApplicationConfiguration.DatacenterAvailability.SIDECAR
+            == context.config.getDatacenterAvailability()) {
```

Comment:
curious, why doesn't `LOCAL` do `beatMetrics()` ?

Reply:
Because LOCAL is used in the setup where you have one Reaper instance in a DC and no Reaper instance in the other DCs, so there's no metric request going on. Reaper will reach out directly to the nodes that are in the local DC and proceed.

Reply:
I'm not sure that's necessarily true. `LOCAL` can also mean that availability and running repairsis preferred (over prevention of node overloading) when all reapers in a datacenter go down.

Reply:
I think considering `LOCAL` for this use case goes against what the documentation states and would complexify both the code and the tests.
I advise on keeping things as described in the docs so that we don't add confusion over what's already not very easy to understand.
I firmly believe that once the `SIDECAR` mode is out we'll want to reduce the `datacenterAvailability` options anyway to further simplify both usage and the code (I know we still disagree for now on this, but... 😉 ).

Reply:
> I advise on keeping things as described in the docs so that we don't add confusion over what's already not very easy to understand.

The docs state for `LOCAL` that `In addition, metrics can be collected asynchronously through the Apache Cassandra storage.`

This matches my understanding. The difference between `LOCAL` and `EACH` should be simplified to how failure (to jmx connect) is handled. To also also include the definition of how/what it does adds complexity.
That is, the simplest way to explain, as I have understood it, is: they work in the same way but `EACH` stops working if no reaper is running in a datacenter to ensure no C* node is overloaded, while `LOCAL` provides better Reaper availability by continuing to run in presuming the C* nodes in a datacenter where no Reaper is running are ok.

Reply:
> I firmly believe that once the SIDECAR mode is out we'll want to reduce the datacenterAvailability options anyway to further simplify both usage and the code.

Yes, I disagree. A cluster of 200 nodes means 100GB and 100+ virtual CPU usages resources dedicated to Reaper.
It is an opinionated approach, presuming that operators can colocate processes. Different platform container systems and setups vary a lot and I don't think we should be making this presumption.
It also breaks the ability for one Reaper UI to register and visualise all your Cassandra clusters in one UI.
These are huge trade-offs on the assumptions jmx connectivity is difficult for everyone. As we have seen many users have no problems with the jmx connectivity, even between datacenters around the world.

Promotion of the side-car mode should be via new functionality and features we add that are only available when using side-car mode.

Reply:
The doc is misleading as it is not how `LOCAL` works : https://github.com/thelastpickle/cassandra-reaper/blob/1.4/src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java#L653

We're only going through getting the metrics from the backend (and requesting them) if we're using `EACH` and it should stay that way. It also matches the multi DC architecture docs: http://cassandra-reaper.io/docs/usage/multi_dc/

Reply:
> The doc is misleading as it is not how LOCAL works

No, that doc has been intentionally written that way. There's different setups and scenarios that validate `LOCAL` in different ways.

From http://cassandra-reaper.io/docs/usage/multi_dc/

> Single Reaper instance with JMX accessible for a single DC only

^that is not accurate. It provides only one possible scenario of when `LOCAL` is desired. Reaper in `LOCAL` mode can also have multiple Reaper instances running in different datacenters.
`LOCAL` is not about definitely not having a Reaper instance, so much as about how to handle what happens when the (or all) Reaper instance(s) in a datacenter stop.

The idea is that it is about the trade-off between protecting Cassandra nodes from repairs that can overload them versus prioritising that repairs run even if some Reaper instances go down.
It allows the operator to decide what is most important in their environment: healthy nodes versus repaired data.
I don't think we can be opinionated about the Cassandra clusters, datacenters can be a mixture of online (OLTP) and offline (OLAP). In OLAP datacenters latency is less of a concern, so node load is less of a concern, while batch jobs depend upon repaired data for their consistency model.

 > how LOCAL works : https://github.com/thelastpickle/cassandra-reaper/blob/1.4/src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java#L653

That was a bug which has been fixed, if you look in master. It was also an important fix for our integration tests in CircleCI that run against multiple datacenters. We had a discussion about how this impacted CircleCI in the past but I'm guessing the relevance to this discussion wasn't obvious then.

Bear in mind it was I that argued for and introduced `LOCAL` as a compliment to `EACH`, and my understanding of the reasoning for it hasn't changed.
I suspect that the metrics requests need to know how to back-off, just like jmx hosts are blacklisted… so to cover the case where `LOCAL` is used where one or more datacenters don't (and won't ever) have a Reaper instance running.

Reply:
There's a discussion to be had here about a) improving `Heart` to do `beatMetrics()` when in `LOCAL` mode,  and b) better documenting what the different advantages `EACH` and `LOCAL` modes offer.

For now, as `Heart` did not already do `beatMetrics()` for `LOCAL` mode, this discussion can be taken out of the PR.